### ISSUE TITLE
Some commands don't path properly

### ISSUE THREAD

**Issue by praetorg** [bedrocklinux/bedrocklinux-userland#99]
An example from a running Ubuntu strata on a hijacked Arch install:

`nano /etc/network/interfaces`

seems to path to `/bedrock/strata/arch/etc/network/interfaces`, which doesn't exist, rather than `/bedrock/strata/ubuntu/etc/network/interfaces` which is the expected behavior as that is the current target of `/etc/network/interfaces`. Attempting to do so while in the current working directory at `/etc/network` also produces the same result. Currently have to use the full path explicitly rather than the virtual path.

**Comment by Mouvedia** [comment]
Do you have nano on Arch as well?

**Comment by praetorg** [comment]
It was installed on the Arch strata and not the Ubuntu strata. Ubuntu strata was just the bootstrap with init and networking installed. Should paths be intercepted and rewritten in this way? I can foresee potential script malfunctions if that should be the case.

**Comment by Mouvedia** [comment]
> in the users current context

This concept doesn't exist AFAIK.

> Is there any way…

https://bedrocklinux.org/0.7/commands.html#brl-strat

**Comment by paradigm** [comment]
Have you read through [the basic usage documentation](https://bedrocklinux.org/0.7/basic-usage.html)?  I think it covers the expectation for this situation at the end.  It's also covered in [the concepts and terminology page](https://bedrocklinux.org/0.7/concepts-and-terminology.html) along with some background on why Bedrock behaves this way.  If you did, and it didn't make sense to you, let me know and we can work on rephrasing it to be more clear.

I *think* you figured it out in the end, but I'll go through and respond to bits to make sure we're both on the same page.

> An example from a running Ubuntu strata on a hijacked Arch install:
> 
> `nano /etc/network/interfaces`
> 
> seems to path to `/bedrock/strata/arch/etc/network/interfaces`, which doesn't exist, rather than `/bedrock/strata/ubuntu/etc/network/interfaces` which is the expected behavior as that is the current target of `/etc/network/interfaces`. Attempting to do so while in the current working directory at `/etc/network` also produces the same result. Currently have to use the full path explicitly rather than the virtual path.

`/etc/network/interfaces` is a *local* path.  This means you have multiple `/etc/network/interfaces` paths: your arch stratum has a `/etc/network/interfaces` and your ubuntu stratum has a `/etc/network/interfaces`.  Bedrock does this because it doesn't know if arch and ubuntu have different preferences for what should be at that file path.

If you try to do something with `/etc/network/interfaces` and don't tell Bedrock which one you want, Bedrock has to figure it out on its own.  By default, all programs that come from the arch stratum see arch's `/etc/network/interfaces` and all programs from ubuntu see ubuntu's `/etc/network/interfaces`.  This means if arch's programs and ubuntu's programs expect something different at `/etc/network/interfaces`, each will work because they'll both see their own version.  You used arch's `nano` to read `/etc/network/interfaces`, and so you saw arch's instance of `/etc/network/interfaces`.

If you want to specify which instance of a local path to use in a given context, you have to prefix `/bedrock/strata/<stratum>` to tell Bedrock which one you want.  Here, you could use `nano /bedrock/strata/arch/etc/network/interfaces`.

Other file paths are *global*, which means you only have one copy and all strata see the same thing.  This includes, for example, `/etc/passwd`.  For those, all processes see the same path and you don't have to manually prefix `/bedrock/strata/<stratum>`.  Those are used for files and directories Bedrock knows most distros have the same expectations and won't conflict.

> Edit: My mistake, this should be as expected as the call should be rewritten (within system paths) to keep applications working in their context

Exactly.  This is what ensures arch's programs and ubuntu's programs both work without conflicting.

> Is there any way to make commands entered manually work in the users current context?

I don't know what you mean by user's context.  When you run `nano /etc/network/interfaces` there's nothing hinting to Bedrock which `/etc/network/interfaces` you want other than the fact that `nano` comes from arch, which is why it picks arch's `/etc/network/interfaces`.

> Or should we just duplicate application installs per strata to attain desired functionality (within system paths) as a matter of course?

You *could* install `nano` from both arch *and* ubuntu, then run:

```
strat arch nano /etc/network/interfaces
```

or

```
strat ubuntu nano /etc/network/interfaces
```

to specify which `/etc/network/interfaces` you want by specifying which stratum is providing nano.  However, I personally prefer

```
nano /bedrock/strata/arch/etc/network/interfaces
```

or

```
nano /bedrock/strata/ubuntu/etc/network/interfaces
```

as it uses less disk and RAM, and in my opinion its more clear about what your goal is (you don't care about which nano you're using, you care about which `/etc/network/interfaces` you're editing).  Both work fine, though, if you prefer one over the other.

> So my current understanding is to only expect merged usage within shared paths (such as /home).

I think that's correct.  You should only expect global paths like `/home` or `/etc/passwd` to always provide your desired instance of a given file without prefixing `/bedrock/strata/<stratum>`, because there's only one copy of such files.  For local paths, you have to specify which you want.

If that makes sense and clears things up for you, feel free to close the issue.  Otherwise, let me know where you're still lost.

**Comment by Mouvedia** [comment]
@paradigm we could add a list populated with preferences for certain applications in `bedrock.conf.`.
e.g. `nano: 'ubuntu'`

For convenience, it would be supported by `strat` using a flag. The first time you used nano you would set the preferred stratum using `strat --lock ubuntu nano /path/foo` and thence you would just type `nano`. How's that?

**Comment by Mouvedia** [comment]
@praetorg check https://github.com/bedrocklinux/bedrocklinux-userland/blob/master/src/slash-bedrock/etc/bedrock.conf#L309

**Comment by paradigm** [comment]
> @paradigm we could add a list populated with preferences for certain applications in `bedrock.conf.`.
> e.g. `nano: 'ubuntu'`

We already support such a feature in `bedrock.conf`.  Look at how `reboot` is handled in there, for example.  It doesn't look like I've documented that on the website - I made a note to do so when I get the chance.

> For convenience, it would be supported by `strat` using a flag. The first time you used nano you would set the preferred stratum using `strat --lock ubuntu nano /path/foo` and thence you would just type `nano`. How's that?

I don't plan on having utilities automatically change bedrock.conf for the foreseeable future.  I do see value in such automated changes, though.  For example, it'd be nice if `brl rename` renamed defences to a given stratum in `bedrock.conf, or if `brl update` could automatically add new features to `bedrock.conf`.  However, there's a risk that a bug in such a feature would corrupt `bedrock.conf` and completely break the system, and I don't think the benefit is worth such a risk at this point in time.  We could go with a `conf.d/*` pattern, but I think that discourages users from reading `bedrock.conf`.

Moreover, I don't think it'd be `strat`'s responsibility to edit configuration.  `strat`'s job is to override `bedrock.conf` in a given instance, not to write to it.  If I were to make a tool that acts as a front-end to `bedrock.conf`, it'd probably be some new `brl` subcommand.  Consider how `git config` is a high level `git` subcommand rather than some `git commit` flag.

> @praetorg check https://github.com/bedrocklinux/bedrocklinux-userland/blob/master/src/slash-bedrock/etc/bedrock.conf#L309

I don't follow why you're bringing this to praetorg's attention.  If he or she were to pin/lock/whatever `nano` to always come from Ubuntu, it'd provide the "wrong" file when `nano /etc/pacman.conf`.  If it were set to always come from Arch, it'd provide the "wrong" file when `nano `/etc/apt/sources.list` is run.  The issue here, I think, is an understanding of what Bedrock is doing and how users are to use it rather than a configuration issue.
